sql-plugin/src/main/scala/com/nvidia/spark/rapids/PrioritySemaphore.scala

```diff
@@ -77,8 +88,10 @@ class PrioritySemaphore[T](val maxPermits: Int)(implicit ordering: Ordering[T])
     try {
       occupiedSlots -= numPermits
       if (waitingQueue.nonEmpty) {
-        val nextThread = waitingQueue.dequeue()
-        nextThread.condition.signal()
+        val nextThread = waitingQueue.head
+        if (canAcquire(nextThread.numPermits)) {
+          nextThread.condition.signal()
+        }
```

Comment:
When waking up, this needs to be a while loop to awaken all threads that have room.  Consider the case where one thread pops 10 permits, and there's 5 single-permit requests in the queue.  We don't want to wake up just one of them.